sdr_uhd.c

```diff
@@ -0,0 +1,208 @@
+    printf("\n");
+}
+
+bool uhdHandleOption(int argc, char **argv, int *jptr) {
+    int j = *jptr;
+    bool more = (j+1 < argc);
+    if (!strcmp(argv[j], "--uhd-device-args") && more) {
+        uhd.device_args = strdup(argv[++j]);
+    } else {
+        return false;
+    }
+
+    *jptr = j;
+    return true;
+}
+
+#define UHD_RUN(x) {if (x) { uhdClose(); return false; }}
+#define UHD_INIT(x) {++uhd.free_func_ptr; UHD_RUN(x)}
+
+bool uhdOpen() {
+    uhd_tune_request_t tune_request = {
+        .target_freq = Modes.freq,
+        .rf_freq_policy = UHD_TUNE_REQUEST_POLICY_AUTO,
+        .dsp_freq_policy = UHD_TUNE_REQUEST_POLICY_AUTO,
+    };
+    uhd_tune_result_t tune_result;
+
+    ++uhd.free_func_ptr;
+    if (!uhd.device_args) {
+        uhd.device_args = strdup("");
+    }
+
+    UHD_INIT(uhd_usrp_make(&uhd.usrp, uhd.device_args))
+    UHD_INIT(uhd_rx_streamer_make(&uhd.rx_streamer))
+    UHD_INIT(uhd_rx_metadata_make(&uhd.md))
+
+    fprintf(stderr, "setting bw, rx rate to %fMsps\n", Modes.sample_rate / 1e6);
+    UHD_RUN(uhd_usrp_set_rx_rate(uhd.usrp, Modes.sample_rate, uhd.channel))
+    UHD_RUN(uhd_usrp_set_rx_bandwidth(uhd.usrp, Modes.sample_rate, uhd.channel))
+    if (Modes.gain != MODES_DEFAULT_GAIN) {
+        fprintf(stderr, "setting rx gain to %f\n", Modes.gain);
+        UHD_RUN(uhd_usrp_set_rx_gain(uhd.usrp, Modes.gain, uhd.channel, ""))
+    }
+    fprintf(stderr, "setting rx freq to %fMHz\n", Modes.freq / 1e6);
+    UHD_RUN(uhd_usrp_set_rx_freq(uhd.usrp, &tune_request, uhd.channel, &tune_result))
+
+    uhd.converter = init_converter(
+        INPUT_SC16Q11, Modes.sample_rate, Modes.dc_filter, &uhd.converter_state);
+
+    return true;
+}
+
+void uhdRun() {
+    uhd_stream_args_t stream_args = {
+        .cpu_format = "sc16",
+        .otw_format = "sc16",
+        .args = "",
+        .channel_list = &uhd.channel,
+        .n_channels = 1
+    };
+
+    uhd_stream_cmd_t stream_cmd = {
+        .stream_mode = UHD_STREAM_MODE_START_CONTINUOUS,
+        .stream_now = true
+    };
+
+    uhd_stream_cmd_t stop_stream_cmd = {
+        .stream_mode = UHD_STREAM_MODE_STOP_CONTINUOUS,
+    };
+
+    size_t num_rx_samps = 0;
+    struct mag_buf *outbuf = NULL;
+
+    if (uhd_usrp_get_rx_stream(uhd.usrp, &stream_args, uhd.rx_streamer)) {
+        fprintf(stderr, "uhd_usrp_get_rx_stream() failed\n");
+        return;
+    }
+
+    fprintf(stderr, "streaming...\n");
+    uhd.sample_data = malloc(MODES_MAG_BUF_SAMPLES * SAMPLE_SIZE);
+    int64_t full_secs;
+    double frac_secs;
+    uhd_rx_metadata_time_spec(uhd.md, &full_secs, &frac_secs);
+    double first_recv_time = full_secs + frac_secs;
+
+    if (uhd_rx_streamer_issue_stream_cmd(uhd.rx_streamer, &stream_cmd)) {
+        fprintf(stderr, "could not start stream\n");
+        return;
+    }
+
+    for (;;) {
+        outbuf = fifo_acquire(0);
+        if (!outbuf) {
+            fprintf(stderr, "could not get outbuf\n");
```

Comment:
This will happen in normal operation -- 

 * when dump1090 is halting, no new buffers will be returned; it's OK to bail out on this case
 * if the demodulator can't keep up, then the FIFO freelist will drain and no buffer may be returned, but this means "drop the data on the floor", not "abandon ship"

I'd suggest implementing a stop lifecycle function that sets a flag somewhere which can be tested in this loop to distinguish the shutdown vs. overload cases.

Reply:
Done.